toolchain/check/handle_class.cpp

```diff
@@ -642,46 +637,84 @@ static auto CheckCompleteAdapterClassType(Context& context,
       {.type_id = context.GetBuiltinType(SemIR::BuiltinInstKind::WitnessType),
        .object_repr_id = adapted_type_id});
 }
+static auto AddStructTypeFields(
+    Context& context,
+    llvm::SmallVector<SemIR::StructTypeField>& struct_type_fields) {
+  for (auto field_decl_id : context.field_decls_stack().PeekArray()) {
+    auto field_decl = context.insts().GetAs<SemIR::FieldDecl>(field_decl_id);
+    field_decl.index =
+        SemIR::ElementIndex{static_cast<int>(struct_type_fields.size())};
+    context.sem_ir().insts().Set(field_decl_id, field_decl);
```

Comment:
Rather than directly setting the instruction, I think it'd be good to add something like `auto Check::Context::ReplaceInstPreservingConstantValue(SemIR::InstId inst_id, SemIR::Inst inst) -> void`, analogous to the existing `ReplaceInstBeforeConstantUse`. For extra safety, maybe we could make it recompute the constant value after the replacement and CHECK-fail if it's different.

Reply:
Sure, done in 89b88ff78420553468f093d965a1d50b36abe02f - at least a best guess at the requested assert/figuring out how the constant evaluation works. (since the constant value of a FieldDecl is currently just the inst_id itself, so it can't change, so I couldn't really think of anything to do to test the assertion works (like I tried modifying the FieldDecl's name_id or type_id, but those aren't part of the constant))